File: src/main/golang/leetcode/editor/cn/[146]LRU Cache.go
package lc

//leetcode submit region begin(Prohibit modification and deletion)
type Node struct {
	Prev, Next *Node
	key, Val   int
}

func NewNode(k, v int) *Node {
	return &Node{key: k, Val: v}
}

type LRUCache struct {
	route          map[int]*Node
	head, tail     *Node
	capacity, size int
}

func (lst *LRUCache) addLast(n *Node) {
	lst.tail.Prev.Next, lst.tail.Prev, n.Prev, n.Next = n, n, lst.tail.Prev, lst.tail
	lst.size += 1
}
func (lst *LRUCache) remove(n *Node) *Node {
	n.Prev.Next, n.Next.Prev = n.Next, n.Prev
	lst.size -= 1
	return n

}
func (lst *LRUCache) makeRecently(n *Node) {
	lst.remove(n)
	lst.addLast(n)
}
func (lst *LRUCache) removeFirst() *Node {
	return lst.remove(lst.head.Next)
}

func Constructor(capacity int) LRUCache {
	lru := LRUCache{
		route:    make(map[int]*Node),
		head:     NewNode(0, 0),
		tail:     NewNode(0, 0),
		capacity: capacity,
	}
	lru.head.Next, lru.tail.Prev = lru.tail, lru.head
	return lru

}

func (this *LRUCache) Get(key int) int {
	if v, ok := this.route[key]; ok {
		this.makeRecently(v)
		return v.Val
	} else {
		return -1
	}
}

func (this *LRUCache) Put(key int, value int) {
	if v, ok := this.route[key]; ok {
		v.Val = value
		this.makeRecently(v)
	} else {
		n := NewNode(key, value)
		this.route[key] = n
		this.addLast(n)
	}
	for this.size > this.capacity { //evict
		n := this.removeFirst()
		delete(this.route, n.key)
	}
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * obj := Constructor(capacity);
 * param_1 := obj.Get(key);
 * obj.Put(key,value);
 */
//leetcode submit region end(Prohibit modification and deletion)

File: src/main/golang/leetcode/editor/cn/[146]LRU Cache.py
# leetcode submit region begin(Prohibit modification and deletion)
class LRUCache:
    class Node:
        def __init__(self, key, val):
            self.key = key
            self.val = val
            self.prev = None
            self.next = None

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.cache = {}
        self.head = None
        self.tail = None

    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1
        node = self.cache[key]
        self.adjustOrder(node)
        return node.val

    def adjustOrder(self, node):
        if node == self.head:
            return
        if node == self.tail:
            self.tail = node.prev
            self.tail.next = None
        else:
            node.prev.next = node.next
            node.next.prev = node.prev
        node.prev = None
        node.next = self.head
        self.head.prev = node
        self.head = node

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            node = self.cache[key]
            node.val = value
            self.adjustOrder(node)
        else:
            if len(self.cache) == self.capacity:
                del self.cache[self.tail.key]
                if self.tail == self.head:
                    self.head = None
                    self.tail = None
                else:
                    self.tail = self.tail.prev
                    self.tail.next = None
            node = LRUCache.Node(key, value)
            self.cache[key] = node
            if self.head:
                node.next = self.head
                self.head.prev = node
                self.head = node
            else:
                self.head = node
                self.tail = node

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)
# leetcode submit region end(Prohibit modification and deletion)
